s3Client has no path-style addressing, so every server-side object-store call fails with UnknownHostException on echno-objects.echno-minio

## Reported

Anand, QA pass on `echno.in`, 28 Aug 2026. ClickUp `86d45v7nt`.

Attaching a file while editing an Issue fails with "Unexpected error occurred", the detail mentioning an `UnknownException` while interacting with a service. The Issue itself saves, then a second toast says "Issue updated, some files failed".

## Root cause

Confirmed from the staging backend log rather than inferred. Three occurrences on 28 Aug from 09:02 UTC, all identical:

```
software.amazon.awssdk.core.exception.SdkClientException: Received an UnknownHostException
when attempting to interact with a service. ...
Caused by: java.net.UnknownHostException: echno-objects.echno-minio
```

`echno-objects` is the bucket. `echno-minio` is the MinIO container. The SDK is putting the bucket in the **host**, virtual-host style, and Docker DNS has no such name. Only `echno-minio` resolves.

The reason is in `common/configuration/SpacesConfig.java`. There are two S3 beans and only one of them was given path-style addressing.

`s3Presigner()`, lines 50-66, is correct, and its comment even says why:

```java
// Path-style addressing is required for MinIO, which puts the bucket in the path rather than the host.
.serviceConfiguration(
        S3Configuration.builder()
                .pathStyleAccessEnabled(true)
                .build()
)
```

`s3Client()`, lines 30-44, has no `serviceConfiguration` at all:

```java
return S3Client.builder()
        .endpointOverride(URI.create(spaces))
        .credentialsProvider(...)
        .region(Region.US_EAST_1)
        .build();
```

So every server-side object-store call, which is the path the Issue edit form still takes, builds `http://echno-objects.echno-minio` and fails to resolve. The presigned browser path works, because the presigner is configured properly. That is why this looks intermittent depending on which screen you are on.

Staging environment, for the record:

```
DIGITAL_OCEAN_SPACES_URI=http://echno-minio:9000
DIGITAL_OCEAN_SPACES_BUCKET_NAME=echno-objects
OBJECT_STORE_PUBLIC_ENDPOINT=https://storage.echno.in
```

The bucket name has no dots, so this is not a TLS or SNI problem. It is purely addressing style.

## Not a duplicate of the object-store issues already open

Both were checked before filing:

- echno-web #301 (closed) is `img-src` and `next/image` naming the retired DigitalOcean bucket. That blocks **display** of a stored file, after a successful upload.
- echno-deployment #19 (open) is the k8s flavour having no public object-store endpoint and no CSP origin.

This one is the compose flavour's **server-side** client, and it is unfixed on `development`. `git log` shows no commit touching `SpacesConfig.java` since the deployed revision `5a16111`, and the last change to that file, `c146599 fix: presign against the public object-store endpoint`, is what added path-style to the presigner and did not add it to the client.

## What a fix involves

1. Add the same `S3Configuration.builder().pathStyleAccessEnabled(true)` to `s3Client()`. That is the fix. It is correct against DigitalOcean Spaces and AWS as well, so it does not need to be conditional.
2. Stop returning a bare "Unknown error occurred" for an object-store failure. A storage outage should read as a storage outage, and `86d45v7nt` asks for exactly that.
3. On the web side, an upload that fails should offer a retry rather than leaving the user with a saved Issue and a lost file.

Worth checking whether anything else in the codebase builds an S3 client without path style before we call this closed.

--- a/src/main/java/org/tornotron/echno_backend/common/configuration/SpacesConfig.java
+++ b/src/main/java/org/tornotron/echno_backend/common/configuration/SpacesConfig.java
@@ -4,6 +4,7 @@ import org.springframework.beans.factory.annotation.Value;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
 import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
 import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.S3Client;
@@ -12,56 +13,84 @@ import software.amazon.awssdk.services.s3.presigner.S3Presigner;
 
 import java.net.URI;
 
+/**
+ * Object-store clients.
+ *
+ * <p>Two clients talk to the same store: {@link S3Client} for server-side calls and
+ * {@link S3Presigner} for URLs handed to the browser. They differ only in the endpoint
+ * they address, because the browser cannot reach the internal one. Everything else,
+ * credentials, region and above all the addressing style, comes from the single
+ * {@link #objectStoreServiceConfiguration()} bean below, so the two clients cannot end
+ * up disagreeing about how a bucket is addressed.
+ */
 @Configuration
 public class SpacesConfig {
 
+    /** Internal endpoint used for server-side calls. */
     @Value("${digital-ocean.uri}")
     private String spaces;
 
+    /** Public endpoint presigned URLs are signed against, since the browser resolves that one. */
     @Value("${digital-ocean.presign-endpoint}")
     private String presignEndpoint;
 
     @Value("${digital-ocean.access-key-id}")
-    private String access_key_id;
+    private String accessKeyId;
 
     @Value("${digital-ocean.secret-access-key}")
-    private String secret_access_key;
+    private String secretAccessKey;
+
+    /**
+     * Where the bucket name goes in a request URL. Path-style puts it in the path
+     * ({@code http://host/bucket/key}); the alternative, virtual-host style, puts it in the
+     * hostname ({@code http://bucket.host/key}).
+     *
+     * <p>Default true, because it is the setting that works on every store we deploy against.
+     * Self-hosted MinIO serves path-style only, so virtual-host style resolves the bucket into
+     * a hostname that does not exist. DigitalOcean Spaces and AWS S3 both still serve path-style,
+     * so the same default is correct there. Set {@code digital-ocean.path-style-access=false}
+     * only for a store that refuses path-style, and be aware that a bucket name containing dots
+     * then breaks TLS certificate matching.
+     */
+    @Value("${digital-ocean.path-style-access:true}")
+    private boolean pathStyleAccess;
+
+    /**
+     * The one addressing decision, shared by every client below. Both beans take it as an
+     * argument rather than building their own, which is what stops them drifting apart:
+     * a client that wants a different style has to change this bean, and changing it moves
+     * both clients together.
+     */
+    @Bean
+    public S3Configuration objectStoreServiceConfiguration() {
+        return S3Configuration.builder()
+                .pathStyleAccessEnabled(pathStyleAccess)
+                .build();
+    }
 
     @Bean
-    public S3Client s3Client() {
+    public S3Client s3Client(S3Configuration objectStoreServiceConfiguration) {
         return S3Client.builder()
                 .endpointOverride(URI.create(spaces))
-                .credentialsProvider(
-                        StaticCredentialsProvider.create(
-                                AwsBasicCredentials.create(
-                                        access_key_id,
-                                        secret_access_key
-                                )
-                        )
-                )
+                .serviceConfiguration(objectStoreServiceConfiguration)
+                .credentialsProvider(objectStoreCredentials())
                 .region(Region.US_EAST_1)
                 .build();
     }
 
-    // Presigned URLs are handed to the browser, so they must be signed against the
-    // public object-store endpoint (which may differ from the internal endpoint the
-    // s3Client above uses for server-side ops). Path-style addressing is required for
-    // MinIO, which puts the bucket in the path rather than the host.
     @Bean
-    public S3Presigner s3Presigner() {
+    public S3Presigner s3Presigner(S3Configuration objectStoreServiceConfiguration) {
         return S3Presigner.builder()
                 .endpointOverride(URI.create(presignEndpoint))
-                .serviceConfiguration(
-                        S3Configuration.builder()
-                                .pathStyleAccessEnabled(true)
-                                .build()
-                )
+                .serviceConfiguration(objectStoreServiceConfiguration)
+                .credentialsProvider(objectStoreCredentials())
                 .region(Region.US_EAST_1)
-                .credentialsProvider(
-                        StaticCredentialsProvider.create(
-                                AwsBasicCredentials.create(access_key_id, secret_access_key)
-                        )
-                )
                 .build();
+    }
+
+    private AwsCredentialsProvider objectStoreCredentials() {
+        return StaticCredentialsProvider.create(
+                AwsBasicCredentials.create(accessKeyId, secretAccessKey)
+        );
     }
 }

--- a/src/main/resources/application.yml
+++ b/src/main/resources/application.yml
@@ -7,6 +7,7 @@ jwt:
 digital-ocean:
   uri: ${DIGITAL_OCEAN_SPACES_URI}
   presign-endpoint: ${OBJECT_STORE_PUBLIC_ENDPOINT:${DIGITAL_OCEAN_SPACES_URI}}
+  path-style-access: ${DIGITAL_OCEAN_SPACES_PATH_STYLE:true}
   access-key-id: ${DIGITAL_OCEAN_SPACES_KEY_ID}
   secret-access-key: ${DIGITAL_OCEAN_SPACES_KEY_SECRET}
   bucket-name: ${DIGITAL_OCEAN_SPACES_BUCKET_NAME}

--- a/src/test/java/org/tornotron/echno_backend/common/configuration/SpacesConfigAddressingTest.java
+++ b/src/test/java/org/tornotron/echno_backend/common/configuration/SpacesConfigAddressingTest.java
@@ -1,0 +1,111 @@
+package org.tornotron.echno_backend.common.configuration;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.test.util.ReflectionTestUtils;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3Configuration;
+import software.amazon.awssdk.services.s3.model.GetUrlRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.presigner.S3Presigner;
+import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;
+
+import java.net.URI;
+import java.time.Duration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Pins how the object-store clients address a bucket.
+ *
+ * <p>The server-side client and the presigner used to be built independently, and only the
+ * presigner was given path-style addressing. The client therefore folded the bucket into the
+ * hostname and every server-side call died on {@code UnknownHostException:
+ * echno-objects.echno-minio}, because MinIO answers to one name and serves buckets from the
+ * path. These tests fail if either client goes back to bucket-in-host, and fail if the two
+ * ever stop agreeing.
+ */
+class SpacesConfigAddressingTest {
+
+    private static final String INTERNAL_ENDPOINT = "http://echno-minio:9000";
+    private static final String PUBLIC_ENDPOINT = "https://storage.echno.test";
+    private static final String BUCKET = "echno-objects";
+    private static final String KEY = "issue/site-photo.png";
+
+    @Test
+    void serverSideClientPutsTheBucketInThePathNotTheHost() {
+        SpacesConfig config = configWith(true);
+
+        try (S3Client client = config.s3Client(config.objectStoreServiceConfiguration())) {
+            URI url = urlFor(client);
+
+            assertThat(url.getHost()).isEqualTo("echno-minio");
+            assertThat(url.getPath()).isEqualTo("/" + BUCKET + "/" + KEY);
+            assertThat(url.getHost()).doesNotContain(BUCKET);
+        }
+    }
+
+    @Test
+    void presignerPutsTheBucketInThePathNotTheHost() {
+        SpacesConfig config = configWith(true);
+
+        try (S3Presigner presigner = config.s3Presigner(config.objectStoreServiceConfiguration())) {
+            URI url = presignedUrlFor(presigner);
+
+            assertThat(url.getHost()).isEqualTo("storage.echno.test");
+            assertThat(url.getPath()).isEqualTo("/" + BUCKET + "/" + KEY);
+        }
+    }
+
+    @Test
+    void bothClientsAddressTheBucketTheSameWay() {
+        SpacesConfig config = configWith(true);
+        S3Configuration shared = config.objectStoreServiceConfiguration();
+
+        try (S3Client client = config.s3Client(shared);
+             S3Presigner presigner = config.s3Presigner(shared)) {
+
+            // Endpoints differ by design, the addressing style must not.
+            assertThat(urlFor(client).getPath()).isEqualTo(presignedUrlFor(presigner).getPath());
+        }
+    }
+
+    @Test
+    void virtualHostStyleIsAvailableForStoresThatRequireIt() {
+        SpacesConfig config = configWith(false);
+        S3Configuration shared = config.objectStoreServiceConfiguration();
+
+        try (S3Client client = config.s3Client(shared);
+             S3Presigner presigner = config.s3Presigner(shared)) {
+
+            // Both flip together, which is the point of the shared configuration.
+            assertThat(urlFor(client).getHost()).isEqualTo(BUCKET + ".echno-minio");
+            assertThat(presignedUrlFor(presigner).getHost()).isEqualTo(BUCKET + ".storage.echno.test");
+        }
+    }
+
+    private static SpacesConfig configWith(boolean pathStyleAccess) {
+        SpacesConfig config = new SpacesConfig();
+        ReflectionTestUtils.setField(config, "spaces", INTERNAL_ENDPOINT);
+        ReflectionTestUtils.setField(config, "presignEndpoint", PUBLIC_ENDPOINT);
+        ReflectionTestUtils.setField(config, "accessKeyId", "test-key-id");
+        ReflectionTestUtils.setField(config, "secretAccessKey", "test-secret-key");
+        ReflectionTestUtils.setField(config, "pathStyleAccess", pathStyleAccess);
+        return config;
+    }
+
+    /** The URL the client would call for an object, resolved offline through the SDK's own rules. */
+    private static URI urlFor(S3Client client) {
+        return URI.create(client.utilities()
+                .getUrl(GetUrlRequest.builder().bucket(BUCKET).key(KEY).build())
+                .toString());
+    }
+
+    private static URI presignedUrlFor(S3Presigner presigner) {
+        return URI.create(presigner.presignPutObject(PutObjectPresignRequest.builder()
+                        .signatureDuration(Duration.ofMinutes(15))
+                        .putObjectRequest(PutObjectRequest.builder().bucket(BUCKET).key(KEY).build())
+                        .build())
+                .url()
+                .toString());
+    }
+}

